### PR TITLE
feat: Implement FlatESLint

### LINKED ISSUE #15661: Change Request: Matching behavior in flat config
### ESLint version

HEAD

### What problem do you want to solve?

Currently, the flat config matching algorithm considers any config object without a `files` key to match all files. For example, this config matches all files:

```
{
    rules: { semi: "error" }
}
```

The problem with this is when unknown file types are passed into ESLint. For example, if `foo.html` is passed on the command line, that technically matches the config above and ESLint will attempt to parse the HTML file as if it were JavaScript. That will, necessarily, result in a parsing error.

This is the way the [RFC](https://github.com/eslint/rfcs/tree/main/designs/2019-config-simplification) intended for config objects to work, but in reality, this behavior is confusing from a user perspective. (Wasn't immediately obvious until I implemented flat config.)

### What do you think is the correct solution?

I think that config objects without `files` should not apply to all files, but instead, they should be applied *only when a file matches another config object's `files` array*. Example:

```
[
    {
        plugins: {
            markdown: require("eslint-plugin-markdown")
        }
    },
    {
        files: ["**/*.js"],
        rules: { semi: "error" }
    },
    {
        files: ["**/*.md"],
        processor: "markdown/markdown"
    }
]
```

In this case, the files `foo.md` and `foo.js` will be considered matched and will inherit the from the first config object. The file `foo.html` will not be considered a match and will be treated as if it were explicitly ignored because it doesn't directly match any `files` patterns.

So config objects without `files` will still apply to any file that is matched elsewhere, they just won't apply on their own.

### Participation

- [X] I am willing to submit a pull request for this change.

### Additional comments

_No response_

### LINKED ISSUE #15683: Change Request: Config base path in flat config
### ESLint version

main

### What problem do you want to solve?

When implementing flat config, I discovered a quirk that wasn't obvious from the RFC. The `files` and `ignores` patterns are calculated relative to where the config file was found. So if the `eslint.config.js` file is found in `~/project`, then `*.js` matches `~/project/foo.js` but not `~/foo.js`. This is the intended behavior and not a problem, as I think it is expected when we search for a config file that the patterns will be relative to where the file was found.

The problem is when someone uses the `ESLint` constructor and passes in an alternate config file (via `configFile` option). In this case, we don't do a search for `eslint.config.js` and instead just load the referenced config file. Right now, that config file also resolves `files` and `ignores` relative to the location of the file, but I believe this is unexpected from a user perspective.

### What do you think is the correct solution?

When someone passes `configFile` to the `ESLint` constructor, I think `files` and `ignores` should be calculated as relative to the current working directory instead of the location of the file. 

### Participation

- [X] I am willing to submit a pull request for this change.

### Additional comments

_No response_

### PR DESCRIPTION
**Pull request by nzakas**
#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[x] Add something to the core
[ ] Other, please explain:

#### What changes did you make? (Give an overview)

This is the first implementation of `FlatESLint`! :tada:

Things to know:

1. `FlatESLint` implements the flat config system. It will eventually replace the current `ESLint` class completely, so long term we will have one `ESLint` class that uses flat config. Just in the interim will we have two of these classes.
2. `FlatESLint` combines all of the functionality of `ESLint` and `CLIEngine`, so we will be able to delete `CLIEngine` once we switch over to the new config system for good.
3. Caching is not yet implemented. That's a bit tricky and rather than holding up everything else, I'll work on that separately.
4. `FlatESLint` and `FlatRuleTester` are exposed from the `use-at-your-own-risk` entrypoint so we can get some developer feedback as we go along.
5. `FlatESLint` is not accessible through the command line. This is essentially a developer-only preview and not an end-user preview.
6. There are a lot fewer tests than with `ESLint`. Some of the tests no longer made sense for `FlatESLint`, while others didn't even make sense for `ESLint` (there are a lot of tests that test the same things). I did my best to make sure the most important tests work.
7. Some tests are marked out by `xdescribe` or `xit` if I was blocked on fixing them, or if they relate to caching. 

Refs #13481
Fixes #15661
Fixes #15683

#### Is there anything you'd like reviewers to focus on?

Please focus on big picture issues. There will probably be some compatibility issues with the existing config system that we'll discover along the way. The goal of this PR is not to be perfect, but to land a first-pass of the new config system that we can build upon before we diverge too far away.

Also, I left a note in `linter.js` about some of the extra checks I need to do to avoid weird errors.

### CODE REVIEW

**Comment by nzakas** [comment]
There are some incompatibilities between the way flat config works in this PR and the way it worked when I did the original work, so I'm needing to go back through the original work and update that to get the tests to pass.

File: lib/linter/linter.js [hunk 1]
```diff
@@ -1770,12 +1775,24 @@ class Linter {
...
+                {
+                    ruleId: null,
+                    severity: 1,
+                    message: `No matching configuration found for ${filename}.`,
+                    line: 0,
+                    column: 0
+                }
+            ];
+        }
+
```
Comment (nzakas) [inline lib/linter/linter.js:1795]: Unlike with eslintrc, it's possible that the filename given might not match something in the config array. If that's the case, I've opted to create a warning saying that the file didn't have a configuration.

I wasn't sure if this should be considered an error, so I used the model of creating a warning to match what we do for ignored files in `ESLint`, but definitely open to other ideas.

**Comment by nzakas** [comment]
After a ferocious battle with Node.js 12 compatibility, everything is working!

File: tests/lib/eslint/flat-eslint.js [hunk 2]
```diff
@@ -0,0 +1,4647 @@
...
+const shell = require("shelljs");
+const hash = require("../../../lib/cli-engine/hash");
+const { unIndent, createCustomTeardown } = require("../../_utils");
+const coreRules = require("../../../lib/rules");
+
+//------------------------------------------------------------------------------
+// Tests
+//------------------------------------------------------------------------------
+
+describe("FlatESLint", () => {
```
Comment (btmills) [inline tests/lib/eslint/flat-eslint.js:31]: I diffed this file against `tests/lib/eslint/flat-eslint.js` in VSCode to review. The omitted/updated/added tests all make sense, understanding that ignoring and caching are still being worked on.

**Review by btmills — APPROVED** [review]
Bravo :clap: I'm not seeing any structural issues here. Especially since this will be behind `use-at-your-own-risk`, I'd rather get this in and iterate than exhaustively review every corner case here. Other than the merge conflict, which I can fix myself if you don't get to it, I don't see why this shouldn't be merged before I do the release tomorrow (well, technically later today).

**Comment by btmills** [comment]
This is behind the `use-at-your-own-risk` export, so we can still experiment freely. Merging to include in this release!

**Pull request merged**